### PR TITLE
Feature/typescript

### PR DESCRIPTION
**Pull request by dror-weiss**
Add typescript support.

This will enable the following from a package using Typescript for development:
```ts
import mergician from 'mergician';
```

Closes: #1

### CODE REVIEW

**Comment by jhildenbiddle** [comment]
Brilliant. Thank you, @dror-weiss. I will get to this sometime next week.

**Comment by Arraying** [comment]
> Brilliant. Thank you, @dror-weiss. I will get to this sometime next week.

Any updates on this?

**Comment by jhildenbiddle** [comment]
My apologies for the delay @dror-weiss. I've been away but will return on Sunday. I will merge this as soon as I get back.

Thanks again!

File: types/mergician.d.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	 * const mergedObj = mergician({
+	 *     appendArrays: true,
+	 *     dedupArrays: true
+	 * })(obj1, obj2, obj3);
+	 *
+	 * console.log(mergedObj); // { a: 1, b: [2, 3], c: { d: 2, e: 3 } }
+	 *
+	 * @param options
+	 */
+	function mergician<T extends ObjectLiteral>(options: MergicianOptions): (...objects: ObjectLiteral[]) => T;
```
Comment (jhildenbiddle) [inline types/mergician.d.ts:111]: FWIW, having the function overloads available that demonstrate different ways of calling `mergician` is great. I wasn't aware of this functionality. I'm going to update the JSDoc comments to match after this PR is merged.

File: package.json [hunk 2]
```diff
@@ -65,7 +68,8 @@
     "prepare": "run-s clean build",
     "serve": "node server.js",
     "start": "run-p serve 'build -- --watch'",
-    "test": "jest",
+    "test": "jest && yarn test:types",
+	"test:types": "tsc -p tests/types",
```
Comment (jhildenbiddle) [inline package.json:72]: Am I correct that the purpose of this test to verify that `mergician.d.ts` will not break the TypeScript compiler? If so, the code contained within `tests/types/index.ts` is irrelevant and could be replaced with something as simple as the the following, yes?

```js
import mergician from 'mergician';

mergician({});
```

**Review by jhildenbiddle — CHANGES_REQUESTED** [review]
Finally getting around to this. I appreciate your patience, @dror-weiss. 

I do have a few questions, most of which I'm sure are due to my lack of familiarity with TypeScript. The two assumptions I'm have about these TypeScript definitions is that they will...

1. Catch type mismatch issues in IDEs (VS Code) and during compilation (`tsc -p`)
2. Display relevant information in IDEs while editing

The issue I'm having is that neither of the two benefits above are working. Details below.

File: types/mergician.d.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+		skipUniversalKeys?: boolean;
+
+		// ------
+		// Arrays
+		// ------
+
+		/**
+		 * Merge array values at the end of existing arrays
+		 */
+		appendArrays?: boolean;
```
Comment (jhildenbiddle) [inline types/mergician.d.ts:49]: If option names and types are defined by the `MergicianOptions` interface, I would expect the following to throw an error:

```js
const customMergeFn = mergician({
  appendArrays: 1
});
```

Unfortunately, no error is displayed in VS Code or when during compilation.

**Comment by jhildenbiddle** [comment]
@dror-weiss -- Ping!
@Arraying -- Feel free to jump in.

**Comment by threeal** [comment]
any progress on this? @jhildenbiddle

**Comment by jhildenbiddle** [comment]
@dror-weiss, @threeal --

Type definitions are now available in v2. Closing this PR.

**Pull request closed without merging**